File: Java/424.java
class Solution {
    public int characterReplacement(String s, int k) {
        int n = s.length();
        int[] freq = new int[26];
        int left = 0, maxCount = 0, maxLength = 0;
        for (int right = 0; right < n; right++) {
            maxCount = Math.max(maxCount, ++freq[s.charAt(right) - 'A']);
            if (right - left + 1 - maxCount > k) {
                freq[s.charAt(left) - 'A']--;
                left++;
            }
            maxLength = Math.max(maxLength, right - left + 1);
        }
        return maxLength;
    }
}

// don't need to update maxCount in the if since it is valid at one point, a new
// window must have the same maxCount or larger for that window to have a chance
// to be larger then the current one.˝

File: Python/424.py
from collections import defaultdict


class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        freq = defaultdict(int)
        max_char_count = 0
        result = 0
        left = 0
        for right in range(len(s)):
            freq[s[right]] += 1
            max_char_count = max(max_char_count, freq[s[right]])

            if right - left + 1 - max_char_count > k:
                freq[s[left]] -= 1
                left += 1

            result = max(result, right - left + 1)
        return result
